mlflow/sklearn/__init__.py

```diff
@@ -884,12 +1153,81 @@ def patched_fit(original, self, *args, **kwargs):
+                    result = fit_mlflow(original, self, *args, **kwargs)
+                status.register_model(self, mlflow.active_run().info.run_id)
+                return result
             else:
                 return original(self, *args, **kwargs)
 
+    def patched_predict(original, self, *args, **kwargs):
+        status = _get_autolog_training_status()
+        if status.should_log_post_training_metrics() and status.get_run_id_for_model(self):
+            predict_result = original(self, *args, **kwargs)
+            eval_dataset = args[0] if len(args) >= 1 else kwargs.get('X')
+            eval_dataset_name = status.register_eval_dataset(self, eval_dataset)
+            status.register_prediction_result(
+                self, eval_dataset_name, predict_result)
+            return predict_result
+        else:
+            return original(self, *args, **kwargs)
+
+    def patched_metric_api(original, *args, **kwargs):
+        status = _get_autolog_training_status()
+        if status.should_log_post_training_metrics():
+            # one metric api may call another metric api,
+            # to avoid this, call disable_log_post_training_metrics to avoid nested patch
+            with status.disable_log_post_training_metrics():
+                metric = original(*args, **kwargs)
+
+            if status.is_metrics_value_loggable(metric):
+                metric_name = original.__name__
+                if metric_name.strip() == '<lambda>':
+                    metric_name = 'unknown_metric'
+
+                call_command = status.gen_metric_call_command(metric_name, *args, **kwargs)
+
+                is_register_ok, run_id, metric_key = \
+                    status.register_metric_api_call(metric_name, call_command, args, kwargs)
+                if is_register_ok:
+                    status.log_eval_metric(run_id, metric_key, metric)
+
+            return metric
+        else:
+            return original(*args, **kwargs)
+
+    # we need patch model.score method because:
+    #  some model.score() implementation won't call metric APIs in `sklearn.metrics`
+    #  e.g.
+    #  https://github.com/scikit-learn/scikit-learn/blob/82df48934eba1df9a1ed3be98aaace8eada59e6e/sklearn/covariance/_empirical_covariance.py#L220
+    def patched_model_score(original, self, *args, **kwargs):
+        status = _get_autolog_training_status()
+        if status.should_log_post_training_metrics() and status.get_run_id_for_model(self):
+            with status.disable_log_post_training_metrics():
+                score_value = original(self, *args, **kwargs)
+
+            if status.is_metrics_value_loggable(score_value):
+                metric_name = f'{self.__class__.__name__}_score'
+                call_fn_name = f'{self.__class__.__name__}.score'
+                call_command = status.gen_metric_call_command(call_fn_name, *args, **kwargs)
+
+                eval_dataset = args[0] if len(args) >= 1 else kwargs.get('X')
```

Comment:
@WeichenXu123 Have you checked the first argument of `score` is always named `X`?

Reply:
Not always X

for model.score, exception case includes:
```
X_test, EmpiricalCovariance
X_test, GraphicalLasso
X_test, GraphicalLassoCV
X_test, LedoitWolf
X_test, MinCovDet
X_test, OAS
X_test, ShrunkCovariance
```
for model.predict, exception case includes:
T, IsotonicRegression

and for `LocalOutlierFactor`, https://scikit-learn.org/stable/modules/generated/sklearn.neighbors.LocalOutlierFactor.html

its `predict` is not an method but an property 😂

Reply:
Can we handle these cases by checking for `X_test`? Can we also handle the case where we can't find an eval_dataset by printing a warning message?